I would say CMake, since according to https://github.com/qt/qtbase/commit/41aff1cd938457539fd6fd1009a3859de6b78095:

```
Building Qt in a symlinked build directory leads to all kind of issues,
like unnecessary rebuilds, strange compilation errors or already errors
at configure time.
```

It sufficed to remove the check for building Qt 6.4.2 (expect Qt WebEngine), but it's not the case for Qt 6.5.0. Fortunately, we can build it under `/nix/store`, which is not symlinked. Good luck to the Homebrew guys...

---

I converted this PR to draft, since this hack is also necessary to build Qt Declarative 6.5.0.

```
error: builder for '/nix/store/lb89cxk9aq9waa41cypa6p4pahy4ds91-qtdeclarative-6.5.0.drv' failed with exit code 1;
       last 10 log lines:
       > In file included from /tmp/nix-build-qtdeclarative-6.5.0.drv-0/qtdeclarative-everywhere-src-6.5.0/src/qml/../3rdparty/masm/config.h:43:
       > In file included from /tmp/nix-build-qtdeclarative-6.5.0.drv-0/qtdeclarative-everywhere-src-6.5.0/src/qml/../3rdparty/masm/wtf/Platform.h:33:
       > In file included from /tmp/nix-build-qtdeclarative-6.5.0.drv-0/qtdeclarative-everywhere-src-6.5.0/build/include/QtQml/6.5.0/QtQml/private/qv4global_p.h:1:
       > In file included from /tmp/nix-build-qtdeclarative-6.5.0.drv-0/qtdeclarative-everywhere-src-6.5.0/build/include/QtQml/6.5.0/QtQml/private/../../../../../../src/qml/jsruntime/qv4global_p.h:22:
       > In file included from /tmp/nix-build-qtdeclarative-6.5.0.drv-0/qtdeclarative-everywhere-src-6.5.0/build/include/QtQml/qtqmlglobal.h:1:
       > /tmp/nix-build-qtdeclarative-6.5.0.drv-0/qtdeclarative-everywhere-src-6.5.0/build/include/QtQml/../../../src/qml/qtqmlglobal.h:9:10: fatal error: 'QtQml/qtqml-config.h' file not found
       > #include <QtQml/qtqml-config.h>
       >          ^~~~~~~~~~~~~~~~~~~~~~
       > 1 error generated.
       > ninja: build stopped: subcommand failed.
       For full logs, run 'nix log /nix/store/lb89cxk9aq9waa41cypa6p4pahy4ds91-qtdeclarative-6.5.0.drv'.
```

--- a/pkgs/build-support/setup-hooks/move-build-tree.sh
+++ b/pkgs/build-support/setup-hooks/move-build-tree.sh
@@ -1,0 +1,12 @@
+prePhases+=" moveBuildDir"
+
+moveBuildDir() {
+    mkdir -p $out/.build
+    cd $out/.build
+}
+
+postPhases+=" removeBuildDir"
+
+removeBuildDir() {
+    rm -rf $out/.build
+}

--- a/pkgs/development/libraries/qt-6/default.nix
+++ b/pkgs/development/libraries/qt-6/default.nix
@@ -38,7 +38,8 @@ let
         withGtk3 = true;
         inherit (srcs.qtbase) src version;
         inherit developerBuild;
-        inherit (darwin.apple_sdk_11_0.frameworks) AGL AVFoundation AppKit GSS MetalKit;
+        inherit (darwin.apple_sdk_11_0.frameworks)
+          AGL AVFoundation AppKit Contacts CoreBluetooth EventKit GSS MetalKit;
         patches = [
           ./patches/0001-qtbase-qmake-always-use-libname-instead-of-absolute-.patch
           ./patches/0002-qtbase-qmake-fix-mkspecs-for-darwin.patch
@@ -46,12 +47,6 @@ let
           ./patches/0004-qtbase-fix-locating-tzdir-on-NixOS.patch
           ./patches/0005-qtbase-deal-with-a-font-face-at-index-0-as-Regular-f.patch
           ./patches/0006-qtbase-qt-cmake-always-use-cmake-from-path.patch
-          # Remove symlink check causing build to bail out and fail.
-          # https://gitlab.kitware.com/cmake/cmake/-/issues/23251
-          (fetchpatch {
-            url = "https://github.com/Homebrew/formula-patches/raw/c363f0edf9e90598d54bc3f4f1bacf95abbda282/qt/qt_internal_check_if_path_has_symlinks.patch";
-            sha256 = "sha256-Gv2L8ymZSbJxcmUijKlT2NnkIB3bVH9D7YSsDX2noTU=";
-          })
         ];
       };
       env = callPackage ./qt-env.nix { };

--- a/pkgs/development/libraries/qt-6/modules/qtbase.nix
+++ b/pkgs/development/libraries/qt-6/modules/qtbase.nix
@@ -71,10 +71,14 @@
 , unixODBC
 , unixODBCDrivers
   # darwin
+, moveBuildTree
 , xcbuild
 , AGL
 , AVFoundation
 , AppKit
+, Contacts
+, CoreBluetooth
+, EventKit
 , GSS
 , MetalKit
   # optional dependencies
@@ -166,6 +170,9 @@ stdenv.mkDerivation rec {
     AGL
     AVFoundation
     AppKit
+    Contacts
+    CoreBluetooth
+    EventKit
     GSS
     MetalKit
   ] ++ lib.optional libGLSupported libGL;
@@ -184,7 +191,8 @@ stdenv.mkDerivation rec {
   ++ lib.optional (libmysqlclient != null) libmysqlclient
   ++ lib.optional (postgresql != null) postgresql;
 
-  nativeBuildInputs = [ bison flex gperf lndir perl pkg-config which cmake xmlstarlet ninja ];
+  nativeBuildInputs = [ bison flex gperf lndir perl pkg-config which cmake xmlstarlet ninja ]
+    ++ lib.optionals stdenv.isDarwin [ moveBuildTree ];
 
   propagatedNativeBuildInputs = [ lndir ];
 

--- a/pkgs/development/libraries/qt-6/qtModule.nix
+++ b/pkgs/development/libraries/qt-6/qtModule.nix
@@ -3,6 +3,7 @@
 , cmake
 , ninja
 , perl
+, moveBuildTree
 , srcs
 , patches ? [ ]
 }:
@@ -19,7 +20,8 @@ stdenv.mkDerivation (args // {
   patches = args.patches or patches.${pname} or [ ];
 
   buildInputs = args.buildInputs or [ ];
-  nativeBuildInputs = (args.nativeBuildInputs or [ ]) ++ [ cmake ninja perl ];
+  nativeBuildInputs = (args.nativeBuildInputs or [ ]) ++ [ cmake ninja perl ]
+    ++ lib.optionals stdenv.isDarwin [ moveBuildTree ];
   propagatedBuildInputs = args.qtInputs ++ (args.propagatedBuildInputs or [ ]);
 
   moveToDev = false;

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -1173,6 +1173,10 @@ with pkgs;
     name = "keep-build-tree-hook";
   } ../build-support/setup-hooks/keep-build-tree.sh;
 
+  moveBuildTree = makeSetupHook {
+    name = "move-build-tree-hook";
+  } ../build-support/setup-hooks/move-build-tree.sh;
+
   enableGCOVInstrumentation = makeSetupHook {
     name = "enable-gcov-instrumentation-hook";
   } ../build-support/setup-hooks/enable-coverage-instrumentation.sh;

